### ISSUE TITLE
Example using optim?

### ISSUE THREAD

**Issue by willfrey** [twitter/torch-distlearn#7]
Can you provide an example of how to use AllReduceSGD with the optim package?

I'm confused about whether or not I should normalize gradParameters like they do in [this example.]https://github.com/torch/tutorials/blob/master/2_supervised/4_train.lua) The behavior of AllReduceSGD.sumAndNormalizeGradients isn't clear to me, specifically whether (1) it sums and normalizes the accumulated gradient across all nodes or (2) if it takes the batch size given to the torch-dataset package into account.

I'm curious about the last part because you normalize the batch size, per node in your cifar example.

Thanks.

**Comment by zakattacktwitter** [comment]
Hi,

We just wrote our own quick version of SGD for the cifar10 example. This block of code is equivalent to calling optim.SGD (https://github.com/twitter/torch-distlearn/blob/master/examples/cifar10.lua#L159). The most basic version of optim.SGD is just a multiply and add.

The batch normalization occurs in AllReduceSGD.sumAndNormalizeGradients as you suspected. This is equivalent to what you see in the torch example here (https://github.com/torch/tutorials/blob/master/2_supervised/4_train.lua#L173).

The reason we do this inside the AllReduceSGD code is because not all nodes in the training set will have the same number of steps every time. For example if you have 11 items to train on, your batch size is 4 and you train on 2 nodes. You would see 3 iterations thru the loop with batch sizes of 4, 4 and 3. The last batch of 3 would have 2 items from the first node and 1 from the second node.

We hide this corner case and handle it for you. Hope this helps!

**Comment by willfrey** [comment]
Thanks for your reply!

I'd like to use Nesterov momentum in my training. It sounds like I'd have
to take care of this myself instead of using optim.sgd to do the update. Is
that correct?

On Fri, Mar 4, 2016 at 10:05 AM, zakattacktwitter <notifications@github.com>
wrote:

> Hi,
>
> We just wrote our own quick version of SGD for the cifar10 example. This
> block of code is equivalent to calling optim.SGD (
> https://github.com/twitter/torch-distlearn/blob/master/examples/cifar10.lua#L159).
> The most basic version of optim.SGD is just a multiply and add.
>
> The batch normalization occurs in AllReduceSGD.sumAndNormalizeGradients as
> you suspected. This is equivalent to what you see in the torch example here
> (
> https://github.com/torch/tutorials/blob/master/2_supervised/4_train.lua#L173
> ).
>
> The reason we do this inside the AllReduceSGD code is because not all
> nodes in the training set will have the same number of steps every time.
> For example if you have 11 items to train on, your batch size is 4 and you
> train on 2 nodes. You would see 3 iterations thru the loop with batch sizes
> of 4, 4 and 3. The last batch of 3 would have 2 items from the first node
> and 1 from the second node.
>
> We hide this corner case and handle it for you. Hope this helps!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/torch-distlearn/issues/7#issuecomment-192314237>
> .
>



-- 
Will Frey
Software Engineer |* Digital Reasoning*
Cell: 703.915.8592 <615.838.0457>
901 N Stuart St., Suite 902
Arlington, VA 22203
www.digitalreasoning.com

**Comment by zakattacktwitter** [comment]
Pretty sure that optim SGD supports momentum, but you’d have to check that
package source and documentation to be absolutely sure.

**Comment by willfrey** [comment]
Oh, I misread your comment. I see, I can use optim.sgd instead of your
quick version. My mistake!

It sounds like the correct way to continue is to compute my gradients as
normal using optim.sgd and leave the normalization to distlearn. As I
understand it, distlearn will take care of both: (1) normalizing the
gradient over each minibatch on a node and (2) normalizing the gradients
from each node with respect to one another, so every node gets the same
gradient update.

Can you address why you adapt the batch size in your CIFAR10 example?

Thank you!

On Fri, Mar 4, 2016 at 2:11 PM, zakattacktwitter <notifications@github.com>
wrote:

> Pretty sure that optim SGD supports momentum, but you’d have to check that
> package source and documentation to be absolutely sure.
>
>
> On Fri, Mar 4, 2016 at 11:09 AM, willfrey <notifications@github.com>
> wrote:
>
> > Thanks for your reply!
> >
> > I'd like to use Nesterov momentum in my training. It sounds like I'd have
> > to take care of this myself instead of using optim.sgd to do the update.
> Is
> > that correct?
> >
> > On Fri, Mar 4, 2016 at 10:05 AM, zakattacktwitter <
> > notifications@github.com>
> > wrote:
> >
> > > Hi,
> > >
> > > We just wrote our own quick version of SGD for the cifar10 example.
> This
> > > block of code is equivalent to calling optim.SGD (
> > >
> >
> https://github.com/twitter/torch-distlearn/blob/master/examples/cifar10.lua#L159
> > ).
> > > The most basic version of optim.SGD is just a multiply and add.
> > >
> > > The batch normalization occurs in AllReduceSGD.sumAndNormalizeGradients
> > as
> > > you suspected. This is equivalent to what you see in the torch example
> > here
> > > (
> > >
> >
> https://github.com/torch/tutorials/blob/master/2_supervised/4_train.lua#L173
> > > ).
> > >
> > > The reason we do this inside the AllReduceSGD code is because not all
> > > nodes in the training set will have the same number of steps every
> time.
> > > For example if you have 11 items to train on, your batch size is 4 and
> > you
> > > train on 2 nodes. You would see 3 iterations thru the loop with batch
> > sizes
> > > of 4, 4 and 3. The last batch of 3 would have 2 items from the first
> node
> > > and 1 from the second node.
> > >
> > > We hide this corner case and handle it for you. Hope this helps!
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> >
> https://github.com/twitter/torch-distlearn/issues/7#issuecomment-192314237
> > >
> > > .
> > >
> >
> >
> >
> > --
> > Will Frey
> > Software Engineer |* Digital Reasoning*
> > Cell: 703.915.8592 <615.838.0457>
> > 901 N Stuart St., Suite 902
> > Arlington, VA 22203
> > www.digitalreasoning.com
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/twitter/torch-distlearn/issues/7#issuecomment-192418537
> >
> > .
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/torch-distlearn/issues/7#issuecomment-192420160>
> .
>



-- 
Will Frey
Software Engineer |* Digital Reasoning*
Cell: 703.915.8592 <615.838.0457>
901 N Stuart St., Suite 902
Arlington, VA 22203
www.digitalreasoning.com

**Comment by zakattacktwitter** [comment]
We adapt the batch size so that all the parameters stay the same regardless of how many nodes you train on. For example, if you train on one node/gpu with a batch size of 64 and get all your other parameters in a good state (learning rate, etc). Then when you run this same setup on 2 nodes/gpus you should still compute a total batch size of 64 (32 on each node/gpu) in order for everything to stay consistent.

The whole algorithm is premised on the distribution of the mini batch computation.

**Comment by willfrey** [comment]
That makes sense.

Thank you! I'll close this.
